File: bilmarknad-mcp/src/bilmarknad_mcp/riddermark.py
from __future__ import annotations

import json
import re
from typing import Any
from urllib.parse import quote, urlencode

import httpx

from bilmarknad_mcp.schema import CarListing
from bilmarknad_mcp.soh import apply_soh

RIDDERMARK_BASE = "https://www.riddermarkbil.se"
SEARCH_PATH = "/kopa-bil/"
_NEXT_DATA_RE = re.compile(
    r'<script id="__NEXT_DATA__"[^>]*>(.*?)</script>',
    re.DOTALL,
)


def _make_slug(make: str) -> str:
    return quote((make or "").strip().lower(), safe="")


def _listing_id(make: str, license_plate: str) -> str:
    return f"{_make_slug(make)}/{license_plate.strip().lower()}"


def _listing_url(make: str, license_plate: str) -> str:
    return f"{RIDDERMARK_BASE}/kopa-bil/{_listing_id(make, license_plate)}/"


def _cover_image(item: dict[str, Any]) -> str | None:
    cover = item.get("coverImage")
    if isinstance(cover, dict) and cover.get("url"):
        return cover["url"]
    images = item.get("images") or []
    if images and isinstance(images[0], dict):
        return images[0].get("url")
    return None


def _location_name(item: dict[str, Any]) -> str | None:
    location = item.get("location") or item.get("physicalLocation") or {}
    if isinstance(location, dict):
        return location.get("name")
    return None


def _parse_car(item: dict[str, Any], *, detail: bool = False) -> CarListing:
    make = item.get("make") or ""
    license_plate = item.get("licenseplate") or ""
    listing = CarListing(
        source="riddermark",
        id=_listing_id(make, license_plate),
        title=item.get("title") or item.get("carName") or "",
        make=make or None,
        model=item.get("model") or None,
        year=item.get("modelYear"),
        mileage_km=item.get("mileage"),
        price_sek=item.get("price"),
        fuel=item.get("fuelType"),
        transmission=item.get("gearboxType"),
        location=_location_name(item),
        dealer_name="Riddermark Bil",
        url=_listing_url(make, license_plate),
        image_url=_cover_image(item),
        published_at=item.get("publishedAt"),
        registration_number=license_plate or None,
        raw=item,
    )
    source = "riddermark_detail" if detail else "riddermark_search"
    apply_soh(
        listing,
        item.get("title"),
        item.get("modelDescription"),
        item.get("fullText"),
        source=source,
    )
    return listing


def _extract_next_data(html: str) -> dict[str, Any]:
    match = _NEXT_DATA_RE.search(html)
    if not match:
        return {}
    return json.loads(match.group(1))


class RiddermarkClient:
    def __init__(self, client: httpx.Client | None = None):
        self._client = client
        self._owns = client is None

    def _get_client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(
                timeout=30.0,
                headers={"User-Agent": "bilmarknad-mcp/0.1"},
                follow_redirects=True,
            )
        return self._client

    def _fetch_next_data(self, path: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        query = f"?{urlencode(params, doseq=True)}" if params else ""
        response = self._get_client().get(f"{RIDDERMARK_BASE}{path}{query}")
        response.raise_for_status()
        return _extract_next_data(response.text)

    def search(
        self,
        q: str | None = None,
        make: str | None = None,
        model: str | None = None,
        price_min: int | None = None,
        price_max: int | None = None,
        mileage_max_km: int | None = None,
        rows: int = 20,
        page: int = 1,
    ) -> list[CarListing]:
        params: dict[str, Any] = {"page": page}
        search_parts = [part for part in (q, make, model) if part]
        if search_parts:
            params["search"] = " ".join(search_parts)
        if price_min is not None:
            params["priceFrom"] = price_min
        if price_max is not None:
            params["priceTo"] = price_max
        if mileage_max_km is not None:
            params["mileageTo"] = mileage_max_km

        data = self._fetch_next_data(SEARCH_PATH, params)
        cars = ((data.get("props") or {}).get("pageProps") or {}).get("carsJson") or []
        listings = [_parse_car(item) for item in cars if isinstance(item, dict)]
        return listings[:rows]

    def get_listing(self, listing_id: str) -> CarListing | None:
        listing_id = (listing_id or "").strip().strip("/")
        if not listing_id:
            return None

        if "/" in listing_id:
            make_slug, license_plate = listing_id.split("/", 1)
            path = f"/kopa-bil/{make_slug}/{license_plate}/"
            data = self._fetch_next_data(path)
            advert = ((data.get("props") or {}).get("pageProps") or {}).get("advertJson")
            if isinstance(advert, dict):
                return _parse_car(advert, detail=True)
            return None

        results = self.search(q=listing_id, rows=5, page=1)
        target = listing_id.lower()
        for item in results:
            if item.registration_number and item.registration_number.lower() == target:
                return self.get_listing(item.id)
            if item.id.split("/", 1)[-1] == target:
                return self.get_listing(item.id)
        return None

    def close(self) -> None:
        if self._owns and self._client is not None:
            self._client.close()


def parse_riddermark_url(url):
    import importlib
    mod=importlib.import_module("bilmarknad_mcp.urls")
    parsed=mod.parse_listing_url(url)
    return parsed if parsed and parsed[0]=="riddermark" else None

File: internal/bilmarknad/riddermark/riddermark.go
package riddermark

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"

	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/httputil"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/schema"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/soh"
)

const baseURL = "https://www.riddermarkbil.se"

var nextDataRE = regexp.MustCompile(`(?s)<script id="__NEXT_DATA__"[^>]*>(.*?)</script>`)

// Client scrapes Riddermark Next.js page data.
type Client struct {
	httpClient *http.Client
	owns       bool
}

func NewClient(c *http.Client) *Client {
	if c == nil {
		c = httputil.NewRedirectClient(map[string]string{"User-Agent": httputil.UserAgent})
	}
	return &Client{httpClient: c, owns: c.Transport == nil}
}

func makeSlug(make string) string {
	return url.PathEscape(strings.ToLower(strings.TrimSpace(make)))
}

func listingID(make, licensePlate string) string {
	return makeSlug(make) + "/" + strings.ToLower(strings.TrimSpace(licensePlate))
}

func listingURL(make, licensePlate string) string {
	return baseURL + "/kopa-bil/" + listingID(make, licensePlate) + "/"
}

func coverImage(item map[string]any) *string {
	if cover, ok := item["coverImage"].(map[string]any); ok {
		if u, ok := cover["url"].(string); ok && u != "" {
			return &u
		}
	}
	if images, ok := item["images"].([]any); ok && len(images) > 0 {
		if img, ok := images[0].(map[string]any); ok {
			if u, ok := img["url"].(string); ok && u != "" {
				return &u
			}
		}
	}
	return nil
}

func locationName(item map[string]any) *string {
	for _, key := range []string{"location", "physicalLocation"} {
		if loc, ok := item[key].(map[string]any); ok {
			if name, ok := loc["name"].(string); ok && name != "" {
				return &name
			}
		}
	}
	return nil
}

func parseCar(item map[string]any, detail bool) schema.CarListing {
	make := fmt.Sprint(item["make"])
	licensePlate := fmt.Sprint(item["licenseplate"])
	dealer := "Riddermark Bil"
	listing := schema.CarListing{
		Source:             "riddermark",
		ID:                 listingID(make, licensePlate),
		Title:              coalesceTitle(item),
		Make:               strPtr(make),
		Model:              strPtr(fmt.Sprint(item["model"])),
		Year:               intPtr(item["modelYear"]),
		MileageKM:          intPtr(item["mileage"]),
		PriceSEK:           intPtr(item["price"]),
		Fuel:               strPtr(fmt.Sprint(item["fuelType"])),
		Transmission:       strPtr(fmt.Sprint(item["gearboxType"])),
		Location:           locationName(item),
		DealerName:         &dealer,
		URL:                schema.StrPtr(listingURL(make, licensePlate)),
		ImageURL:           coverImage(item),
		PublishedAt:        strPtr(fmt.Sprint(item["publishedAt"])),
		RegistrationNumber: strPtr(licensePlate),
		Raw:                item,
	}
	source := "riddermark_search"
	if detail {
		source = "riddermark_detail"
	}
	soh.Apply(&listing, source,
		fmt.Sprint(item["title"]),
		fmt.Sprint(item["modelDescription"]),
		fmt.Sprint(item["fullText"]),
	)
	return listing
}

func coalesceTitle(item map[string]any) string {
	if t := fmt.Sprint(item["title"]); t != "" && t != "<nil>" {
		return t
	}
	return fmt.Sprint(item["carName"])
}

func extractNextData(html string) map[string]any {
	m := nextDataRE.FindStringSubmatch(html)
	if len(m) < 2 {
		return map[string]any{}
	}
	var data map[string]any
	_ = json.Unmarshal([]byte(m[1]), &data)
	return data
}

func (c *Client) fetchNextData(ctx context.Context, path string, params url.Values) (map[string]any, error) {
	u := baseURL + path
	if len(params) > 0 {
		u += "?" + params.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("riddermark: %s", resp.Status)
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	return extractNextData(string(body)), nil
}

func (c *Client) Search(ctx context.Context, q, makeName, model *string, priceMin, priceMax, mileageMax *int, rows, page int) ([]schema.CarListing, error) {
	params := url.Values{}
	params.Set("page", fmt.Sprint(page))
	var parts []string
	for _, p := range []*string{q, makeName, model} {
		if p != nil && *p != "" {
			parts = append(parts, *p)
		}
	}
	if len(parts) > 0 {
		params.Set("search", strings.Join(parts, " "))
	}
	if priceMin != nil {
		params.Set("priceFrom", fmt.Sprint(*priceMin))
	}
	if priceMax != nil {
		params.Set("priceTo", fmt.Sprint(*priceMax))
	}
	if mileageMax != nil {
		params.Set("mileageTo", fmt.Sprint(*mileageMax))
	}
	data, err := c.fetchNextData(ctx, "/kopa-bil/", params)
	if err != nil {
		return nil, err
	}
	cars := digCars(data)
	out := make([]schema.CarListing, 0, len(cars))
	for _, item := range cars {
		out = append(out, parseCar(item, false))
	}
	if rows > 0 && len(out) > rows {
		out = out[:rows]
	}
	return out, nil
}

func digCars(data map[string]any) []map[string]any {
	props, _ := data["props"].(map[string]any)
	pageProps, _ := props["pageProps"].(map[string]any)
	cars, _ := pageProps["carsJson"].([]any)
	out := make([]map[string]any, 0, len(cars))
	for _, car := range cars {
		if m, ok := car.(map[string]any); ok {
			out = append(out, m)
		}
	}
	return out
}

func (c *Client) GetListing(ctx context.Context, listingID string) (*schema.CarListing, error) {
	listingID = strings.Trim(strings.TrimSpace(listingID), "/")
	if listingID == "" {
		return nil, nil
	}
	if strings.Contains(listingID, "/") {
		parts := strings.SplitN(listingID, "/", 2)
		path := fmt.Sprintf("/kopa-bil/%s/%s/", parts[0], parts[1])
		data, err := c.fetchNextData(ctx, path, nil)
		if err != nil {
			return nil, err
		}
		props, _ := data["props"].(map[string]any)
		pageProps, _ := props["pageProps"].(map[string]any)
		if advert, ok := pageProps["advertJson"].(map[string]any); ok {
			l := parseCar(advert, true)
			return &l, nil
		}
		return nil, nil
	}
	results, err := c.Search(ctx, &listingID, nil, nil, nil, nil, nil, 5, 1)
	if err != nil {
		return nil, err
	}
	target := strings.ToLower(listingID)
	for _, item := range results {
		if item.RegistrationNumber != nil && strings.ToLower(*item.RegistrationNumber) == target {
			return c.GetListing(ctx, item.ID)
		}
		parts := strings.SplitN(item.ID, "/", 2)
		if len(parts) == 2 && parts[1] == target {
			return c.GetListing(ctx, item.ID)
		}
	}
	return nil, nil
}

func (c *Client) Close() {
	if c.owns && c.httpClient != nil {
		c.httpClient.CloseIdleConnections()
	}
}

func strPtr(s string) *string {
	s = strings.TrimSpace(s)
	if s == "" || s == "<nil>" {
		return nil
	}
	return &s
}

func intPtr(v any) *int {
	switch n := v.(type) {
	case float64:
		i := int(n)
		return &i
	case int:
		return &n
	default:
		var i int
		if _, err := fmt.Sscan(fmt.Sprint(v), &i); err != nil {
			return nil
		}
		return &i
	}
}
